drivers/mpu9250/mpu9250.c

```diff
@@ -0,0 +1,209 @@
+#define MEM_OFFSET_DATA_MAG_Y (0x1C)
+#define MEM_OFFSET_DATA_MAG_Z (0x20)
+#define MEM_OFFSET_TIMESTAMP_LOW (0x24)
+#define MEM_OFFSET_TIMESTAMP_HIGH (0x28)
+
+typedef struct
+{
+  uint32_t timestamp_lo;
+  uint32_t timestamp_hi;
+  uint16_t gyro_x;
+  uint16_t gyro_y;
+  uint16_t gyro_z;
+  uint16_t acc_x;
+  uint16_t acc_y;
+  uint16_t acc_z;
+  uint16_t mag_x;
+  uint16_t mag_y;
+  uint16_t mag_z;
+} buffer_t;
+
+struct data
+{
+  void *regs;
+  buffer_t buffer;
+  int size;
+  struct miscdevice misc;
+};
+
+/*
+ * @brief This function gets executed on fread.
+ */
+static int dev_read(struct file *filep, char *buf, size_t count,
+                    loff_t *offp)
+{
+  struct data *dev = container_of(filep->private_data,
+                                  struct data, misc);
+  unsigned int rdata;
+  buffer_t tmpBuf;
+
+  if (BUF_SIZE != sizeof(tmpBuf))
+  {
+    printk(KERN_ERR "Data struct buffer_t is not allocated as expected.\n");
+    BUG();
+    return -ENOEXEC;
+  }
+
+  /* check out of bound access */
+  if ((*offp < 0) || (*offp >= BUF_SIZE))
+    return 0;
+
+  /* limit number of readable bytes to maximum which is still possible */
+  if ((*offp + count) > BUF_SIZE)
+    count = BUF_SIZE - *offp;
+
+  /* read data from FPGA and store into kernel space buffer */
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_GYRO_X);
+  tmpBuf.gyro_x = (rdata & 0x0000FFFF);
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_GYRO_Y);
+  tmpBuf.gyro_y = (rdata & 0x0000FFFF);
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_GYRO_Z);
+  tmpBuf.gyro_z = (rdata & 0x0000FFFF);
+
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_ACC_X);
+  tmpBuf.acc_x = (rdata & 0x0000FFFF);
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_ACC_Y);
+  tmpBuf.acc_y = (rdata & 0x0000FFFF);
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_ACC_Z);
+  (tmpBuf.acc_z = (rdata & 0x0000FFFF));
+
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_MAG_X);
+  tmpBuf.mag_x = (rdata & 0x0000FFFF);
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_MAG_Y);
+  tmpBuf.mag_y = (rdata & 0x0000FFFF);
+  rdata = ioread16(dev->regs + MEM_OFFSET_DATA_MAG_Z);
+  tmpBuf.mag_z = (rdata & 0x0000FFFF);
+
+  tmpBuf.timestamp_lo = ioread32(dev->regs + MEM_OFFSET_TIMESTAMP_LOW);
+  tmpBuf.timestamp_hi = ioread32(dev->regs + MEM_OFFSET_TIMESTAMP_HIGH);
+
+  /* ------ TESTING ONLY --------------------------------------------------------- */
+  /* PROVIDE DUMMY DATA */
+  tmpBuf.gyro_x = 0x1111;
+  tmpBuf.gyro_y = 0x2222;
+  tmpBuf.gyro_z = 0x3333;
+  tmpBuf.acc_x = 0x4444;
+  tmpBuf.acc_y = 0x5555;
+  tmpBuf.acc_z = 0x6666;
+  tmpBuf.mag_x = 0x7777;
+  tmpBuf.mag_y = 0x8888;
+  tmpBuf.mag_z = 0x9999;
+  tmpBuf.timestamp_lo = 0x47001100;
+  tmpBuf.timestamp_hi = 0x48001200;
+  /* ------ TESTING ONLY --------------------------------------------------------- */
+
+  /* copy data from kernel space buffer into user space */
+  if (count > 0)
+    count = count - copy_to_user(buf, (char *)&dev->buffer + *offp, count);
```

Comment:
not sure whether the part with `(char*)&dev->buffer` is correct.. :thinking:

Reply:
as long as `dev->buffer` is the struct, and not a pointer to it, you should be fine